Updates for DOCS 464, 442 and 415 (#450)
clerkinc/clerk-docs: docs/users/sync-data.mdx

---
title: Sync Clerk data to your backend with webhooks
description: The recommended way to sync data between Clerk and your application's backend is via webhooks. In this guide you'll learn how to enable webhooks and how to set up your backend so that it is updated every time an event happens on your Clerk instance.
---

import { Tab, Tabs, Steps, Callout } from 'nextra-theme-docs';

> import { CodeBlockTabs } from '@/components/CodeBlockTabs';

Author: Like nuke them all?

> import { CodeBlockTabs } from '@/components/CodeBlockTabs';

Reviewer: Yup!

To retrieve your Webhook Signing Secret, click on the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) page in the side nav of the [Clerk Dashboard](https://dashboard.clerk.com). 
<Images
  width={2000}
  height={1125}
  src="/images/users/guides/sync-backend/sign-secret.png"
  alt="The Webhooks page in the Clerk Dashboard. There is a red arrow pointing to where the Signing Secret is located."
/>
You will need to set this value as an environment variable in your project. This guide uses `WEBHOOK_SECRET` as the key. However, you can set the key to whatever you like; just be sure to update the code examples.
```env filename=".env.local"
WEBHOOK_SECRET=your_signing_secret
```

### Understanding the webhook payload

The Clerk webhook events are sent as HTTP `POST` requests with a JSON body. All messages contain:
- `data` - an object that holds information for the event's payload.
- `object` - this is always `event`
- `type` - the type of webhook event. See [Supported webhook events](/docs/integrations/webhooks#supported-webhook-events) for a full list.

Below is an example of a webhook object with no payload:

```json
{
  "data": {

>     // The event type specific payload will be here.

Reviewer 2: Invalid JSON. We don't allow it (see STYLEGUIDE) because if folks don't remove the comments after pasting, it can bugger things up. Any other way to do this?

Author: I don't know the best place to land here. The concern about copy and pasting is definitely legitimate, but shouldn't apply here. The developer shouldn't be copying this -- its just a stand in. 

I was trying to avoid using a specific example but instead trying to keep this agnostic. If we wanted a full example, I would say to go with a *.deleted one, such as the `user.deleted` sample:
```js
{
  "data": {
    "deleted": true,
    "id": "user_29wBMCtzATuFJut8jO2VNTVekS4",
    "object": "user"
  },
  "object": "event",
  "type": "user.deleted"
}
```

We definitely want something different that the current one which is valid JSON but is also a very incomplete (and maybe misleading) version of user.created.

```js
{
  "data": {
    "birthday": "",
    "created_at": 1654012591514,
    "email_addresses": [
      {
        "email_address": "james@clerk.com"
      }
    ]
  },
  "object": "event",
  "type": "user.created"
}
```

The 'Test the webhook' expands on test the webhook, but will also show all of the sample JSON objects. https://docs-preview-450.clerkpreview.com/docs/users/sync-data#test-the-webhook
We have more information on the payload at https://clerk.com/docs/integrations/webhooks#payload-structure

Reviewer 2: Pity we can't do something like...
```json
{
  "data": {
    "birthday": "",
    "created_at": 1654012591514,
    "email_addresses": [
      {
        "email_address": "james@clerk.com"
      }
    ]
},
   [event_payload](https://clerk.com/docs/integrations/webhooks#payload-structure)
},
```

I'm not going to block this because it hasn't become a problem before, but let's start thinking about better ways to document JSON objects

You'll be presented with a form where you can specify the URL of your backend endpoint. This is the URL where Clerk will send the webhook events. You can also specify the events you want to receive. For example, if you only want to receive events related to users, you can select the **user** option.

> <Callout>
> If you are developing on your localhost, you will need to expose your endpoint to the internet to work with webhooks. See the [Testing the webhook](#testing-the-webhook) section for information on doing this.
> </Callout>
> 

Reviewer: ```suggestion
<Callout>
If you are developing on your localhost, you will need to expose your endpoint to the internet to work with webhooks. See the [Testing the webhook](#testing-the-webhook) section for information.
</Callout>

```

![Add endpoint form](/docs/images/users/guides/sync-backend/webhook-form.png)

Once you click the **Create** button, you'll be presented with your webhook endpoint dashboard. Here you can see the URL of your endpoint and the events you selected and you can also test your endpoint.

> ### Signing secret

Reviewer 2: Can you change this to be action-oriented? Like "Set up your signing secret". (As per our styleguide!)

Author: Changed to "Adding your Signing Secret to your .env.local file" but question about this. In the Dashboard is 'Signing Secret'. The style for the docs is 'Signing secret'. Where should I land on this?

1. Adding your Signing Secret to your .env.local file
2. Adding your 'Signing Secret' to your .env.local file
3. Adding your Signing aecret to your .env.local file
4. or something else? Currently its number 1.

Reviewer: if we call it Signing Secret in the Clerk Dashboard, let's update the entire page to use Signing Secret all throughout. Don't forget to make `.env.local` a code snippet

> ### Signing Secret

Reviewer: Sentence-case headings.

```suggestion
### Signing secret
```

> You will want to retrieve the Webhook signing secret from the Clerk Dashboard. You can find this in the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) section of the Clerk Dashboard. You will need to set this as an environment variable in your project. This code examples below use `WEBHOOK_SECRET` so you can use that, or use what you like and adjust the code examples.
> 
> ![signing-secret-example](/docs/images/users/guides/sync-backend/sign-secret.png)

Reviewer: ```suggestion
To retrieve your Webhook signing secret, click on the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) page in the side nav of the [Clerk Dashboard](https://dashboard.clerk.com). 

<Images
  width={2000}
  height={1125}
  src="/images/users/guides/sync-backend/sign-secret.png"
  alt="The Webhooks page in the Clerk Dashboard. There is a red arrow pointing to where the signing secret is located."
/>

You will need to set this value as an environment variable in your project. This guide uses `WEBHOOK_SECRET` as the key. However, you can set the key to whatever you like; just be sure to update the code examples.

```env filename=".env.local"
WEBHOOK_SECRET=your_signing_secret
```

```

### Understanding the webhook payload

> The Clerk webhook events are sent as HTTP POST requests with a JSON body. All messages contain a `type` string, `id` string, and a `data` object, like this:

Reviewer: Where is the `id` string ? 
I see the keys `data`, `object`, and `type`

> The Clerk webhook events are sent as HTTP POST requests with a JSON body. All messages contain a `type` string, `id` string, and a `data` object, like this:

Reviewer: ```suggestion
The Clerk webhook events are sent as HTTP `POST` requests with a JSON body. All messages contain:

- `data` - an object that holds information for the event's payload ... blah blah
- `object` - blah blah
- `type` - the type of webhook event ? idk whatever needs to be written here

Below is an example of a webhook payload for a `user.created` event:
```

To learn more about the payload structure, check out the [webhooks reference](/docs/integrations/webhooks).

### Install the `Svix` package

> To get started setting up your endpoint, you will need to install the [`svix` package](https://www.npmjs.com/package/svix). Svix provides a package for verifying the webhook signature, making it easy to verify the authenticity of the webhook events.

Reviewer: ```suggestion
To get started setting up your endpoint, you will need to install the [`Svix` package](https://www.npmjs.com/package/svix). Svix provides a package for verifying the webhook signature, making it easy to verify the authenticity of the webhook events.
```

Author: Should this be capitalized? The package is called `svix` where as the service is called `Svix`. I treated this like `@clerk/next`.

Reviewer: good point - I went to the npm page to see what they did, but if its the package (in a code snippet), then yes I agree to not capitalize it!

      return new Response('Error occured -- no svix headers', {
         status: 400
      })
   }

   // Initiate Svix
   const wh = new Webhook(WEBHOOK_SECRET)

   let evt: WebhookEvent

   // Attempt to verify the incoming webhook
   // If successful, the payload will be available from 'evt'
   // If the verification fails, error out and  return error code
   try {
      evt = wh.verify(payload, {
         "svix-id": svix_id,
         "svix-timestamp": svix_timestamp,
         "svix-signature": svix_signature,
      }) as WebhookEvent
   } catch (err: any) {
      // Console log and return errro
      console.log('Webhook failed to verify. Error:', err.message)
      return res.status(400).json({
         success: false,
         message: err.message
      })
   }

   // Grab the ID and TYPE of the Webhook
   const { id } = evt.data;
   const eventType = evt.type;

   console.log(`Webhook with an ID of ${id} and type of ${eventType}`)
   // Console log the full payload to view
   console.log('Webhook body:', evt.data)

   return res.status(200).json({
      success: true,
      message: 'Webhook received'
   })
})
```

</Tab>

</Tabs>

> ### Adding your endpoint to Middleware

Reviewer 2: Use "Add..." instead of "Adding" as per styleguide

Your Route Handler must be made public or ignored by Middleware to allow the request to succeed. The following example will make any webhooks created under `app/api/webhooks/` public. See [authMiddleware](/docs/references/nextjs/auth-middleware) for more information.

 ```tsx filename="middleware.tsx" {4}
 import { authMiddleware } from "@clerk/nextjs";
 
export default authMiddleware({
    publicRoutes: ["/api/webhooks(.*)"]
    });
 
export const config = {
  matcher: ['/((?!.+\\.[\\w]+$|_next).*)', '/', '/(api|trpc)(.*)'],
};
    ```

> ### Testing the webhook

Reviewer 2: Should be "Test" not "testing"

To test a webhook locally, you can use a tool like [ngrok](https://ngrok.com/) or [localtunnel](https://localtunnel.github.io/www/) to expose your local server to the internet. You can then use the ngrok URL as the webhook URL in the Clerk Dashboard.

> If you are testing webhooks outside of your development, you can use our webhook test requests to test your webhook. You can find these in the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) section of the Clerk Dashboard.

Author: I added an image at /images/users/guides/sync-backend/testing-webhooks.png if you could add arrows/boxes? Then I can add it here.

> If you are testing webhooks outside of your development, you can use our webhook test requests to test your webhook. You can find these in the [Webhooks](https://dashboard.clerk.com/last-active?path=webhooks) section of the Clerk Dashboard.

Author: Also updated the messaging here.
